library/std/src/sys/pal/wasip2/net.rs

```diff
@@ -0,0 +1,540 @@
+#![deny(unsafe_op_in_unsafe_fn)]
+
+use self::netc::{c_int, c_void, size_t};
+use super::fd::WasiFd;
+use crate::ffi::CStr;
+use crate::io::{self, BorrowedBuf, BorrowedCursor, IoSlice, IoSliceMut};
+use crate::net::{Shutdown, SocketAddr};
+use crate::os::wasi::io::{AsFd, AsRawFd, BorrowedFd, FromRawFd, IntoRawFd, RawFd};
+use crate::sys::unsupported;
+use crate::sys_common::net::{getsockopt, setsockopt, sockaddr_to_addr, TcpListener};
+use crate::sys_common::{AsInner, FromInner, IntoInner};
+use crate::time::{Duration, Instant};
+use crate::{cmp, mem, str};
+
+#[allow(non_camel_case_types)]
+pub type wrlen_t = size_t;
+
+#[doc(hidden)]
+pub trait IsMinusOne {
+    fn is_minus_one(&self) -> bool;
+}
+
+macro_rules! impl_is_minus_one {
+    ($($t:ident)*) => ($(impl IsMinusOne for $t {
+        fn is_minus_one(&self) -> bool {
+            *self == -1
+        }
+    })*)
+}
+
+impl_is_minus_one! { i8 i16 i32 i64 isize }
+
+pub fn cvt<T: IsMinusOne>(t: T) -> crate::io::Result<T> {
+    if t.is_minus_one() { Err(crate::io::Error::last_os_error()) } else { Ok(t) }
+}
+
+pub fn cvt_r<T, F>(mut f: F) -> crate::io::Result<T>
+where
+    T: IsMinusOne,
+    F: FnMut() -> T,
+{
+    loop {
+        match cvt(f()) {
+            Err(ref e) if e.is_interrupted() => {}
+            other => return other,
+        }
+    }
+}
+
+pub fn cvt_gai(err: c_int) -> io::Result<()> {
+    if err == 0 {
+        return Ok(());
+    }
+
+    if err == netc::EAI_SYSTEM {
+        return Err(io::Error::last_os_error());
+    }
+
+    let detail = unsafe {
+        str::from_utf8(CStr::from_ptr(netc::gai_strerror(err)).to_bytes()).unwrap().to_owned()
+    };
+
+    Err(io::Error::new(
+        io::ErrorKind::Uncategorized,
+        &format!("failed to lookup address information: {detail}")[..],
+    ))
+}
+
+pub fn init() {}
+
+pub struct Socket(WasiFd);
```

Comment:
The WASIp2 target is moving in a different direction than WASIp1 so one refactoring I've been meaning to do for awhile is to reduce the reliance on `WasiFd` in the WASIp2 target. To that effect could you switch this to directly using a `c_int` internally? That would require reproduction of the destruction and removal of the `AsRef<WasiFd>` implementation but that accurately reflects how a socket works on this platform.

Reply:
Hm ok if it's difficult to do then it's fine to leave this in for now. The WASIp2 traits in this regard need to be improved and is something I should put on my plate of things-to-refactor for the future. My guess is that it's probably too much to take on in this PR, so feel free to defer it.

Reply:
If that's easy enough yeah `OwnedFd` I think would be a better fit since we need to eventually sunset `WasiFd` on the p2 target anyway. Again though feel free to defer to a future PR. The main thing I was worried about which was `std::os::wasi` is already unstable on the p2 target, as desired, so what I was originally worried about is already taken care of.

Reply:
Alright, let's defer it then to not block this PR any further.